src/seqme/models/hyformer.py

```diff
@@ -0,0 +1,191 @@
+            for batch in tqdm(
+                _dataloader,
+                disable=not self.verbose,
+            ):
+                batch = batch.to_device(self.device)
+                output = self.model(**batch, return_loss=False)
+                batch_embeddings = output["embeddings"][:, _CLS_TOKEN_IDX].detach().cpu().numpy()
+                embeddings.append(batch_embeddings)
+        return np.concatenate(embeddings, axis=0)
+
+    def compute_perplexity(self, sequences: list[str]) -> np.ndarray:
+        """
+        Compute perplexity for a list of sequences.
+
+        Args:
+            sequences: List of sequences.
+
+        Returns:
+            np.ndarray: Perplexity scores, in the same order as the input sequences.
+        """
+        _TASKS = {"lm": 1.0}
+        _dataloader = self._create_dataloader_fn(
+            dataset=sequences,
+            tasks=_TASKS,
+            tokenizer=self.tokenizer,
+            batch_size=min(len(sequences), self.batch_size),
+            shuffle=False,
+        )
+
+        logit_batches: list[torch.Tensor] = []
+        label_batches: list[torch.Tensor] = []
+
+        with torch.inference_mode():
+            for batch in tqdm(
+                _dataloader,
+                disable=not self.verbose,
+            ):
+                batch = batch.to_device(self.device)
+                output = self.model(**batch, return_loss=False)
+                logit_batches.append(output["logits"].cpu())
+                label_batches.append(batch["input_labels"].cpu())
+
+        logits = torch.cat(logit_batches, dim=0)
+        labels = torch.cat(label_batches, dim=0)
+
+        return self._perplexity_from_logits(logits, labels)
+
+    @staticmethod
+    def _perplexity_from_logits(logits: torch.Tensor, labels: torch.Tensor, ignore_index: int = -100) -> np.ndarray:
+        """Compute sequence-level perplexity from token logits.
+
+        Args:
+            logits: Float tensor of shape (batch, seq_len, vocab_size) with unnormalized scores.
+            labels: Long tensor of shape (batch, seq_len) with token ids used as targets.
+            ignore_index: Index to ignore in the labels.
+
+        Returns:
+            Array of shape (batch,) with perplexity per sequence.
+        """
+        if logits.ndim != 3:
+            raise ValueError("logits must have shape (batch, seq_len, vocab_size)")
+        if labels.ndim != 2:
+            raise ValueError("labels must have shape (batch, seq_len)")
+        if labels.shape[:2] != logits.shape[:2]:
+            raise ValueError("labels and logits must share (batch, seq_len)")
+
+        # log-softmax over the vocabulary for numerical stability
+        log_probs = torch.log_softmax(logits, dim=-1)  # (batch, seq_len, vocab)
+
+        # shift logits and labels by one
+        logits = logits[:, :-1, :].contiguous()
+        labels = labels[:, 1:].contiguous()
+
+        ppls = torch.zeros(logits.shape[0])
```

Comment:
@RasmusML not optimal, but explicit. You can also review, whether it is making what it is suppose to do.